app/post_metadata.go

```diff
@@ -383,74 +417,154 @@ func (a *App) getImagesInMessageAttachments(post *model.Post) []string {
 	return images
 }
 
-func (a *App) getLinkMetadata(requestURL string, timestamp int64, isNewPost bool) (*opengraph.OpenGraph, *model.PostImage, error) {
+func looksLikeAPermalink(url, siteURL string) bool {
+	expression := fmt.Sprintf(`^(%s).*(/pl/)[a-z0-9]{26}$`, siteURL)
+	matched, err := regexp.MatchString(expression, strings.TrimSpace(url))
+	if err != nil {
+		mlog.Warn("error matching regex", mlog.Err(err))
+	}
+	return matched
+}
+
+func postIDFromPermalink(permalinkURL string) string {
+	return permalinkURL[len(permalinkURL)-26:]
+}
+
+func (a *App) getLinkMetadata(requestURL string, timestamp int64, isNewPost bool) (*opengraph.OpenGraph, *model.PostImage, *model.Permalink, error) {
 	requestURL = resolveMetadataURL(requestURL, a.GetSiteURL())
 
 	timestamp = model.FloorToNearestHour(timestamp)
 
 	// Check cache
-	og, image, ok := getLinkMetadataFromCache(requestURL, timestamp)
+	og, image, permalink, ok := getLinkMetadataFromCache(requestURL, timestamp)
+	if !*a.Config().ServiceSettings.EnablePermalinkPreviews {
+		permalink = nil
+	}
+
 	if ok {
-		return og, image, nil
+		return og, image, permalink, nil
 	}
 
 	// Check the database if this isn't a new post. If it is a new post and the data is cached, it should be in memory.
 	if !isNewPost {
-		og, image, ok = a.getLinkMetadataFromDatabase(requestURL, timestamp)
+		og, image, permalink, ok = a.getLinkMetadataFromDatabase(requestURL, timestamp)
+
+		if !*a.Config().ServiceSettings.EnablePermalinkPreviews {
+			permalink = nil
+		}
+
+		if permalink != nil && permalink.PreviewPost == nil {
```

Comment:
Since we're not persisting any information to the DB, should we maybe just skip this step? This branch also duplicates most of the code from below.

Reply:
This populates the permalink that eventually is returned as part of the JSON payload, I'm not sure how this could be skipped. Maybe I'm not seeing something?

Reply:
Maybe I'm misreading this, but the logic looks somewhat like
```
if the metadata is in the database:
    if the url is a permalink:
        load the post and return it

load the post and return it
```
so the information stored in the database isn't being used for permalinks because we're not actually retrieving anything about the preview post with `getLinkMetadataFromDatabase`. That's what's confusing me a bit about this code. I don't think it'll cause any problems, but the logic seems weird to me.